README.md
# Pact Broker Exporter

Prometheus exporter for [Pact Broker](https://github.com/pact-foundation/pact_broker).

[![Docker Pulls](https://img.shields.io/docker/pulls/containersol/pactbroker_exporter.svg)](https://hub.docker.com/r/containersol/pactbroker_exporter/tags)

## Quick Start

This package is available for Docker:

1. Example Pact Broker setup you can find in [pperzyna/pact-example](https://github.com/pperzyna/pact-example) repository.

2. Run Pact Broker Exporter

```bash
docker run --net=host -e PB_EXPORTER_PACTBROKER_URI="http://localhost:9292" containersol/pactbroker_exporter
```

## Building and running

The default way to build is:

```bash
go get github.com/ContainerSolutions/pactbroker_exporter
cd ${GOPATH-$HOME/go}/src/github.com/ContainerSolutions/pactbroker_exporter/
go build -o pactbroker_exporter
export PB_EXPORTER_PACTBROKER_URI="http://localhost:9292"
./pactbroker_exporter <flags>
```

### Flags

* `pactbroker.uri`
  Address of Pact Broker. Default is `http://localhost:9292`.

* `pactbroker.timeout`
  Timeout request to Pact Broker. Default is `5s`.

* `web.listen-address`
  Address to listen on for web interface and telemetry. Default is `:9624`.

* `web.telemetry-path`
  Path under which to expose metrics. Default is `/metrics`.

* `log.level`
  Set logging level: one of `debug`, `info`, `warn`, `error`, `fatal`

* `log.format`
  Set the log output target and format. e.g. `logger:syslog?appname=bob&local=7` or `logger:stdout?json=true`
  Defaults to `logger:stderr`.

### Environment Variables

The following environment variables configure the exporter:

* `PB_EXPORTER_PACTBROKER_URI`
  Address of Pact Broker. Default is `http://localhost:9292`.

* `PB_EXPORTER_PACTBROKER_TIMEOUT`
  Timeout reqeust to Pact Broker. Default is `5s`.

* `PB_EXPORTER_WEB_LISTEN_ADDRESS`
  Address to listen on for web interface and telemetry. Default is `:9624`.

* `PB_EXPORTER_WEB_TELEMETRY_PATH`
  Path under which to expose metrics. Default is `/metrics`.

Settings set by environment variables starting with `PB_` will be overwritten by the corresponding CLI flag if given.


CHANGELOG.md
# CHANGELOG

## 0.1.1 / 2021-02-27

* Migrate to GitHub Actions

## 0.1.0 / 2019-08-14

Initial release


src/pactbroker_exporter.go
package main

import (
	"encoding/json"
	"flag"
	"io/ioutil"
	"log"
	"net/http"
	"os"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

// Flags
var (
	listenAddress  string
	metricPath     string
	dataSourceName string
)

// Metric name parts.
const (
	// Namespace for all metrics.
	namespace = "pactbroker"
)

// Metrics variables
var (
	pactBrokerUp = promauto.NewGauge(prometheus.GaugeOpts{
		Namespace: namespace,
		Name:      "up",
		Help:      "Whether the last scrape of metrics from Pact Broker was able to connect to the server (1 for yes, 0 for no).",
	})
	pactBrokerPacticipants = promauto.NewCounter(prometheus.CounterOpts{
		Namespace: namespace,
		Name:      "pacticipants_total",
		Help:      "The total number of pacticipants",
	})
)

func init() {
	flag.StringVar(&listenAddress, "web.listen-address", getEnv("PB_EXPORTER_WEB_LISTEN_ADDRESS", ":9623"), "Address to listen on for web interface and telemetry.")
	flag.StringVar(&metricPath, "web.telemetry-path", getEnv("PB_EXPORTER_WEB_TELEMETRY_PATH", "/metrics"), "Path under which to expose metrics.")
	flag.StringVar(&dataSourceName, "data-source-name", getEnv("DATA_SOURCE_NAME", "http://localhost:9292"), "Address of Pact Broker")
}

func getEnv(key, defaultValue string) string {
	value := os.Getenv(key)
	if len(value) == 0 {
		return defaultValue
	}
	return value
}

func checkPacticipants() (int, error) {
	r, e := http.Get(dataSourceName + "/pacticipants")
	if e != nil {
		return 0, e
	}
	defer r.Body.Close()

	var result map[string]interface{}

	body, e := ioutil.ReadAll(r.Body)
	if e != nil {
		return 0, e
	}

	json.Unmarshal([]byte(body), &result)

	return len(result["pacticipants"].([]interface{})), nil
}

func main() {
	flag.Parse()

	pactBrokerUp.Set(1)
	pacticipants, e := checkPacticipants()
	if e != nil {
		log.Fatal(e)
	}

	for i := 0; i < pacticipants; i++ {
		pactBrokerPacticipants.Inc()
	}

	http.Handle(metricPath, promhttp.Handler())
	log.Print("Starting Server: ", listenAddress)
	http.ListenAndServe(listenAddress, nil)

}


main.go
package main

import (
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"net/url"
	"sync"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/common/log"
	"github.com/prometheus/common/version"
	"gopkg.in/alecthomas/kingpin.v2"
)

const (
	namespace = "pactbroker"
)

// Exporter structure
type Exporter struct {
	uri   string
	mutex sync.RWMutex
	fetch func(endpoint string) (io.ReadCloser, error)

	pactBrokerUp, pactBrokerPacticipants prometheus.Gauge
	pactBrokerPacts                      *prometheus.GaugeVec
	totalScrapes                         prometheus.Counter
}

// NewExporter function
func NewExporter(uri string, timeout time.Duration) (*Exporter, error) {
	u, err := url.Parse(uri)
	if err != nil {
		return nil, err
	}

	var fetch func(endpoint string) (io.ReadCloser, error)
	switch u.Scheme {
	case "http", "https", "file":
		fetch = fetchHTTP(uri, timeout)
	default:
		return nil, fmt.Errorf("unsupported scheme: %q", u.Scheme)
	}

	return &Exporter{
		uri:   uri,
		fetch: fetch,
		pactBrokerUp: prometheus.NewGauge(
			prometheus.GaugeOpts{
				Namespace: namespace,
				Name:      "up",
				Help:      "The current health status of the server (1 = UP, 0 = DOWN).",
			},
		),
		pactBrokerPacts: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Namespace: namespace,
				Name:      "pacts",
				Help:      "The current number of pacts per pacticipant.",
			},
			[]string{"name"},
		),
		pactBrokerPacticipants: prometheus.NewGauge(
			prometheus.GaugeOpts{
				Namespace: namespace,
				Name:      "pacticipants",
				Help:      "The current number of pacticipants.",
			},
		),
		totalScrapes: prometheus.NewCounter(
			prometheus.CounterOpts{
				Namespace: namespace,
				Name:      "total_scrapes",
				Help:      "The total number of scrapes.",
			},
		),
	}, nil
}

// Describe function of Exporter
func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
	e.mutex.Lock()
	defer e.mutex.Unlock()

	ch <- e.totalScrapes.Desc()
	ch <- e.pactBrokerPacticipants.Desc()
	ch <- e.pactBrokerUp.Desc()
	e.pactBrokerPacts.Describe(ch)
}

// Collect function of Exporter
func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
	e.mutex.Lock()
	defer e.mutex.Unlock()

	up := e.scrape(ch)

	ch <- prometheus.MustNewConstMetric(e.pactBrokerUp.Desc(), prometheus.GaugeValue, up)
	e.pactBrokerPacts.Collect(ch)
}

type pacticipants struct {
	Embedded struct {
		Pacticipants []struct {
			Name string `json:"name"`
		} `json:"pacticipants"`
	} `json:"_embedded"`
}

type pacts struct {
	Links struct {
		Pbpacts []struct {
			Name string `json:"name"`
		} `json:"pb:pacts"`
	} `json:"_links"`
}

func (e *Exporter) scrape(ch chan<- prometheus.Metric) (up float64) {
	e.totalScrapes.Inc()

	var p pacticipants

	body, err := e.fetch("/pacticipants")
	if err != nil {
		log.Errorf("Can't scrape Pack: %v", err)
		return 0
	}
	defer body.Close()

	bodyAll, err := ioutil.ReadAll(body)
	if err != nil {
		return 0
	}

	_ = json.Unmarshal([]byte(bodyAll), &p)

	ch <- prometheus.MustNewConstMetric(e.pactBrokerPacticipants.Desc(), prometheus.GaugeValue, float64(len(p.Embedded.Pacticipants)))

	for _, pacticipant := range p.Embedded.Pacticipants {
		var bodyPact io.ReadCloser
		var pacts pacts

		bodyPact, err = e.fetch("/pacts/provider/" + pacticipant.Name)
		if err != nil {
			log.Errorf("Can't scrape Pack: %v", err)
			return 0
		}
		defer bodyPact.Close()

		pactsAll, err := ioutil.ReadAll(bodyPact)
		if err != nil {
			return 0
		}
		_ = json.Unmarshal([]byte(pactsAll), &pacts)

		e.pactBrokerPacts.WithLabelValues(pacticipant.Name).Set(float64(len(pacts.Links.Pbpacts)))
	}

	return 1
}

func fetchHTTP(uri string, timeout time.Duration) func(endpoint string) (io.ReadCloser, error) {
	tr := &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}
	client := http.Client{
		Timeout:   timeout,
		Transport: tr,
	}

	return func(endpoint string) (io.ReadCloser, error) {
		resp, err := client.Get(uri + endpoint)
		if err != nil {
			return nil, err
		}
		if !(resp.StatusCode >= 200 && resp.StatusCode < 300) {
			resp.Body.Close()
			return nil, fmt.Errorf("HTTP status %d", resp.StatusCode)
		}
		return resp.Body, nil
	}
}

func main() {

	var (
		listenAddress = kingpin.Flag("web.listen-address", "Address to listen on for web interface and telemetry.").Default(":9624").Envar("PB_EXPORTER_WEB_LISTEN_ADDRESS").String()
		metricsPath   = kingpin.Flag("web.telemetry-path", "Path under which to expose metrics.").Default("/metrics").Envar("PB_EXPORTER_WEB_TELEMETRY_PATH").String()
		uri           = kingpin.Flag("pactbroker.uri", "URI of Pact Broker.").Default("http://localhost:9292").Envar("PB_EXPORTER_PACTBROKER_URI").String()
		timeout       = kingpin.Flag("pactbroker.timeout", "Scrape timeout").Default("5s").Envar("PB_EXPORTER_PACTBROKER_TIMEOUT").Duration()
	)

	log.AddFlags(kingpin.CommandLine)
	kingpin.Version(version.Print("pactbroker_exporter"))
	kingpin.HelpFlag.Short('h')
	kingpin.Parse()

	log.Infoln("Starting pactbroker_exporter", version.Info())
	log.Infoln("Build context", version.BuildContext())

	exporter, err := NewExporter(*uri, *timeout)
	if err != nil {
		log.Fatal(err)
	}
	prometheus.MustRegister(exporter)
	prometheus.MustRegister(version.NewCollector("pactexporter"))

	http.Handle(*metricsPath, promhttp.Handler())
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte(`<html><head><title>Pact Broker Exporter</title></head><body><h1>Pact Broker Exporter</h1><p><a href='` + *metricsPath + `'>Metrics</a></p></body></html>`))
	})

	log.Infoln("Listening on", *listenAddress)
	log.Fatal(http.ListenAndServe(*listenAddress, nil))
}
